### ISSUE TITLE
DNU restore fails when APPDATA is not set

### ISSUE THREAD

**Issue by ericstj** [aspnet/dnx#2165]
https://github.com/aspnet/dnx/blob/549c7753dd6fd13a693dc66e3ef8dd1e8effecf3/src/Microsoft.Framework.PackageManager/NuGet/Core/Configuration/Settings.cs#L112 will throw:

```
  System.ArgumentNullException: Value cannot be null.
  Parameter name: path1
     at System.IO.Path.Combine(String path1, String path2, String path3)
     at NuGet.Settings.LoadDefaultSettings(IFileSystem fileSystem, String configFileName, IMachineWideSettings machineWideSettings)
     at Microsoft.Framework.PackageManager.SettingsUtils.ReadSettings(String solutionDir, String nugetConfigFile, IFileSystem fileSystem, IMachineWideSettings machineWideSettings)
     at Microsoft.Framework.PackageManager.RestoreCommand.ReadSettings(String solutionDirectory)
     at Microsoft.Framework.PackageManager.RestoreCommand.<Execute>d__76.MoveNext()
```

This is happening on our build servers where they are running build in a stripped down environment.

**Comment by BrennanConroy** [comment]
I'm guessing you aren't on Windows then, this is a known issue, you can see https://github.com/aspnet/dnx/issues/1813

**Comment by ericstj** [comment]
This is Windows.  It's happening on DevDiv's internal team build servers that are used for official Dev14 builds.

**Comment by BrennanConroy** [comment]
@ChengTian

**Comment by ericstj** [comment]
Simple repro is the following, with no nuget.config in scope.
```
set APPDATA=
dnu restore
```

**Comment by BrennanConroy** [comment]
Makes sense, we should add some checking for environment variable validity. @ChengTian this should probably go in the fix you're doing for correct-local-appdata

**Comment by davidfowl** [comment]
We need to look at what we store in app data. I assume you still have a LocalAppData? We need to cache things somewhere (http cache).

**Comment by ericstj** [comment]
I presume, I'll know tomorrow :smile:   The build rerun process didn't exhibit the issue so I couldn't get any more data.

**Comment by ericstj** [comment]
Build is now hitting this on ProgramData: https://github.com/aspnet/dnx/blob/32358ec9b8970070ff77cbc5beef28c13e3ab09b/src/Microsoft.Framework.PackageManager/NuGet/CommandLine/CommandLineMachineWideSettings.cs#L23.
```

  System.ArgumentNullException: Value cannot be null.
  Parameter name: root
     at NuGet.PhysicalFileSystem..ctor(String root)
     at NuGet.CommandLineMachineWideSettings.<>c__DisplayClass1_0.<.ctor>b__0()
     at System.Lazy`1.CreateValue()
     at System.Lazy`1.LazyInitValue()
     at System.Lazy`1.get_Value()
     at NuGet.CommandLineMachineWideSettings.get_Settings()
     at NuGet.Settings.LoadDefaultSettings(IFileSystem fileSystem, String configFileName, IMachineWideSettings machineWideSettings)
     at Microsoft.Framework.PackageManager.SettingsUtils.ReadSettings(String solutionDir, String nugetConfigFile, IFileSystem fileSystem, IMachineWideSettings machineWideSettings)
     at Microsoft.Framework.PackageManager.RestoreCommand.ReadSettings(String solutionDirectory)
     at Microsoft.Framework.PackageManager.RestoreCommand.<Execute>d__76.MoveNext()
```

**Comment by davidfowl** [comment]
I'm not sure what you're asking for is workable. It sounds like you want a way to override anywhere nuget decides to write stuff. Is that correct?

**Comment by ericstj** [comment]
The only two it hit were ProgramData and APPDATA.  I ended up reproing by starting with an environment completely empty of variables except for PATH, WINDIR, SYSTEM and a few others.  These were the only two that caused a crash and I was able to workaround it by setting them to point to an empty temp directory.

**Comment by ericstj** [comment]
Also, the places that were crashing seemed to only be reading from these paths and already handled missing content.

**Comment by ChengTian** [comment]
So, before `Environment.SpecialFolder.ApplicationData` is available on CoreCLR, we can hardcode some paths when `APPDATA` is not available.

@BrennanConroy , one more thing depending on your PR https://github.com/aspnet/dnx/pull/1692

**Comment by ericstj** [comment]
Don't hardcode.  Since you are just using this to try and find some additional configuration and that configuration is optional just ignore it if you can't resolve the environment variable.

**Comment by davidfowl** [comment]
@ericstj Additional configuration seems similar to setting the right environment variables. Lets say we called them `DNU_CACHE_LOCATION` etc. You'd still need to set them. We write to some of those locations, others are easy to skip.

**Comment by ericstj** [comment]
So I ran a test where I cleared out the entire environment and then ran dnx with appdata and programdata pointing to some temporary folders, specifying the --packages folder explicitly.

I saw it wrote an empty xml file to %APPDATA%\NuGet\NuGet.Config and wrote some .db files to %ProgramData%\Microsoft\Windows\Caches.  The latter seems to be a side-effect of shellexecute.  The former seems under your control and pretty useless.
